rfc/0009-audit-checks.md

```diff
@@ -35,67 +35,50 @@ As a Kubernetes developer, I want access to the audit checks results to allow me
-line arguments as well as a environment variables.
-
-The policy server will not run multiple audit checks at the same time.
-This means that the next audit check should be scheduled just after the current one is finished.
-
-The policy server running in audit mode will be deployed using our Helm charts.
-
-This is a simple flowchart to show an overview how a policy server will work:
-
-```mermaid
-flowchart TD
-  id0(Policy Server start) --> id1{Is in audit mode?}
-  id1 -->|Yes|id7{Are there policies with audit enabled?}
-  id7 -->|Yes|id2(Set timer to trigger next audit check)
-  id2 -->id4(Wait audit check timer run off)
-  id4 -->id5(Run audit checks)
-  id5 -->id6(Report resources violating policies)
-  id6 -->id2
-  id1 -->|No|id3B(Evaluate admission requests)
-  id7 -->|No|id2b(Nothing to do)
-```
+`PolicyServer` needs to store successful evaluation from all policies, and also failed evaluations for 
+policies in monitor mode. Failed evaluations for policies in enforce mode will not be stored as resources won't be 
+created in the cluster. These evaluations will be stored in a `PolicyReport`. There is one `PolicyReport` per namespace and 
+one `ClusterPolicyReport` for cluster-wide resources, see this [RFC](https://github.com/kubewarden/rfc/pull/13/files) for more info.
+
+Many more calls will be made to `PolicyServer` for the background checks, therefore we need to implement a cache
+for policy evaluations. The result of each policy evaluation will be stored in the cache, it will contain the `resourceVersion`
+of the policy and the resource to see if any of them have changed. Before evaluating a new resource it will check in the 
+cache if evaluation was previously done. 
+
+## Background Audit Scanner
+
+By default, background audit scanning happens every hour, and it can be configured via the `background-scan` flag.
+It will scan all resources in the cluster for all policies.
+When a background audit has finished processing all evaluations it will replace all `PolicyReports` and the `ClusterPolicyReport`
+with the new content. This will remove old entries with resources that are deleted and make sure the reports are up to date. 
 
-An audit checks consist of the following steps:
+Background Audit Scanner won't perform policy evaluation, it will make a request to `PolicyServer` for evaluating the resources
+instead.
+
+It should implement a caching mechanism to avoid overloading the Kubernetes api server. We can use client-go caching.
+
+A background audit checks consist of the following steps:
 1. Get information about which policies have audit checks enabled
 2. Get the resources configured in the policies with audit checks enabled
-3. Send the resources that a policy is configured to validate to it and get the validation results.
+3. Request the policy evaluation for all resources to the `PolicyServer`
 4. Aggregate all the policies validation results
```

Comment:
I can't think of an easy to compile a list of resources that are being policed, to serve them to policies.

I think we should just compile a list of all resources, and create bogus requests for those resources, marking those requests as "audit" requests, and serving them directly to the PolicyServers instead of the Kube API. Then, receive all evaluations from their audit policies.
PolicyServers are the ones aware of which things can be evaluated: what kind of audit requests (not those for UPDATE ops for example) and audit policies (not those context-aware, or just looking for UPDATE ops for example).

Reply:
>  PolicyServers are the ones aware of which things can be evaluated: what kind of audit requests (not those for UPDATE ops for example) and audit policies (not those context-aware, or just looking for UPDATE ops for example).

I don't believe PolicyServers are aware of which things can be evaluated. PolicyServers expose a endpoint per policy, then pass the request to the wasm module, however they don't know which resources can be evaluated. This information is inside the rules in the policy spec. That's why I think we should do the following steps:

1. Get information about which policies have audit checks enabled
2. Get the resources configured in the policies with audit checks enabled
3. Request the policy evaluation for all resources to the `PolicyServer`

Reply:
Maybe I'm misunderstanding, as 2 and 3 seem exclusive to me (either approach 2 or 3).
I still can't think of an easy way to compile the list in 2. How do you obtain a list of resources for which a policy applies to?

Isn't it easier to do it the other way around, as in number 3, get the list of all resources, and naively try evaluation of their audit requests? This would mean to send all requests to all PolicyServers, because, how would one know which PolicyServer has which (Cluster)AdmissionPolicy that is relevant?

I would prefer to use dry-run and use the Kube API webhooks to take care of all of this.

Reply:
You can see all resources that will be evaluated for a policy if you look at the `spec.rule` field. For example, in the [psp-capabilities](https://docs.kubewarden.io/quick-start) in the quick start just `Pods` will be evaluated.

Problems I see with kube api dry-run:
1. How do you know which policy rejected the request? We might be able to workaround this by looking at the error message, but doesn't seem like a nice approach.
2. It's not possible to know if a resource would be rejected for more than one policy, as it will be rejected for the one that first does the evaluation.
3. How do we skip evaluation for policies with background-scan=false? I don't see a way to do it with this.
4. What about policies in monitor mode? They will not reject any resources.

Reply:
> You can see all resources that will be evaluated for a policy if you look at the `spec.rule` field. For example, in the [psp-capabilities](https://docs.kubewarden.io/quick-start) in the quick start just `Pods` will be evaluated.

> Problems I see with kube api dry-run:
> 
>     1. How do you know which policy rejected the request? We might be able to workaround this by looking at the error message, but doesn't seem like a nice approach.

If the Audit Scanner is the one creating the dry-run requests, the PolicyServers could be the ones reporting back the results to the Audit Scanner, and they would know the policy.

>     2. It's not possible to know if a resource would be rejected for more than one policy, as it will be rejected for the one that first does the evaluation.

This is a serious shortcoming of dry-run, and possibly a stopper.
 
>     3. How do we skip evaluation for policies with background-scan=false? I don't see a way to do it with this.

I suppose that dry-run requests get marked somehow (I can only see the `?dryRun` in the query) and go through the admission phases, yet Kube API knows to now persist the results into etcd.  It may be the same for us?
Dry-run asks for [custom operators to mark themselves as `sideEffects` of `None` or `NoneOnDryRun`](https://kubernetes.io/docs/reference/using-api/api-concepts/#make-a-dry-run-request). This may tell that custom operators can know that a request if of type dry-run then.

>     4. What about policies in monitor mode? They will not reject any resources.

If the PolicyServers are the ones creating results (by creating PolicyReportResults or by hitting an Rest API), then they have this information.

Reply:
I have done some research on how dry-run works:
1. If you try to create an existing resource with dry-run, it will try to update the resource instead of creating. Which would result in wrong auditing data, since a policy evaluating creations wouldn't run. As a workaround we could change the name, that would trigger a creation event (however I don't like this)
2. AdmissionRequests has a dryRun parameter, so we know in the policy-server if the evaluation is a dry-run
3. There is no way to know which policy rejected a resource If we do the dry-run request from the Audit Scanner. The PolicyServer knows, then it would need to request the Audit Scanner to store the result. This is inconvenient because:
  - we need an extra call
  - what happens if Audit Scanner is not available? we lose data
  - there is no way to aggregate results. We would need to add PolicyReportResult one by one (inefficient).
  - when do we do the cleanup? when we start the background scanner? how do we know it has ended if request are make async from policy-server?
 4. We don't really care about the result of the dry-run call, since Audit Scanner would be waiting for PolicyServer requests. How do we know it has ended if request are make async from policy-server?

Reply:
> I have done some research on how dry-run works:
> 
>     1. If you try to create an existing resource with dry-run, it will try to update the resource instead of creating. 
>         Which would result in wrong auditing data, since a policy evaluating creations wouldn't run. As a
>         workaround we could change the name, that would trigger a creation event (however I don't like this)

Could it be that you are doing a `PUT` (create or update) instead of a `POST` (create)? If using `kubect`, I suppose it does `PUT`.
I've tried it here using `kubectl create --raw` (that just takes the endpoint and does the REST API call), it works as expected and doesn't update the object:
```console
$ kubectl create --raw '/api/v1/namespaces/default/pods?dryRun=All' -f simple-pod.json
Error from server (AlreadyExists): pods "nginx" already exists
```
(tried the same with `kubectl proxy --port=8080 &` and a `curl -X POST`, but I'm failing to call it with the correct `?dryRun=All` query correctly)


> 3\. There is no way to know which policy rejected a resource If we do the dry-run request from the Audit Scanner. The PolicyServer knows, then it would need to request the Audit Scanner to store the result. This is inconvenient because:
> 
> 
>     * we need an extra call

I think we will need extra calls. We need to collect the results from all available PolicyServers. I'm not sure that naively retrying on all the PolicyServers is good enough. And I believe in the future we will need endpoints and a rest API anyways to expose realtime results on policy evaluation.

>     * what happens if Audit Scanner is not available? we lose data

I Audit Scanner is not available, we may not be creating the audit requests in the first place. I do like that if something breaks havoc in the cluster, I can start and stop the audit scanner, freeze the results in time by not letting accept more updates, etc. Even make it scalable.

>     * there is no way to aggregate results. We would need to add PolicyReportResult one by one (inefficient).

We could have implement a simple append log on the Audit Scanner (should be the default). We could use Kafka also if needed. In fact, if we see that we create too many PolicyReportResults because the cluster is busy, it could be configured to not persist them to etcd, but to serve them from a db. Not that we want to implement that in the first run, but the design allows for it.

>     * when do we do the cleanup? when we start the background scanner?

The cleanup is a duty of the Audit Scanner. The start of the background scan (creating the audit requests, and submitting them) is also duty of the Audit Scanner. Both could be configured as needed, be staggered, etc.

>  how do we know it has ended if request are make async from policy-server?

I understood it as the Audit Scanner being the ones creating and POSTing the requests, and naively waiting for PolicyServers to report back results. Each PolicyServer will return X results, one per policy that applies, for each 1 request. We have Y PolicyServers, so X*Y results per 1 requests.


----

Another point on dry-run:

5. Custom controllers need to behave correctly, don't generate side effects if run on dryRun, and explicilty advertise `sideEffects = None` or `NoneOnDryRun`. So far, I can't find one that doesn't do. Still they could be falsely self-advertising. In case they declare sideEffects, the request would not be created IIRC. Which means those resources would not be audited.

Reply:
> Could it be that you are doing a PUT (create or update) instead of a POST (create)? If using kubectl, I suppose it does PUT.
> I've tried it here using kubectl create --raw (that just takes the endpoint and does the REST API call), it works as expected and doesn't update the object:

yes, you are right :) 

> I think we will need extra calls. We need to collect the results from all available PolicyServers. I'm not sure that naively retrying on all the PolicyServers is good enough. And I believe in the future we will need endpoints and a rest API anyways to expose realtime results on policy evaluation

That depends if we will want to store real time results directly from PolicyServer. Having a rest API have the problem of the auditor being offline and we lose data. Also having multiple auditors for HA is a challenge because we need to sync which one start the background check

> We could implement a simple queue on the Audit Scanner (should be the default). We could use Kafka also if needed. In fact, if we see that we create too many PolicyReportResults because the cluster is busy, it could be configured to not persist them to etcd, but to serve them from a db. Not that we want to implement that in the first run, but the design allows for it.

It is not straightforward to create a queue because everything is async and we don't know when it has finished. Communication is like Auditor -> kube api (using dry-run) -> (it might reach PolicyServer or not we don't know) -> back to Auditor (if  PolicyServer was reached). Kafka would be great, but I think it is way to big and uses a lot of resources.

> I think this simplifies saving and deleting PolicyReportResults. Instead of having X PolicyServers in charge of creating/deleting results from a single DB (PolicyReport), we can have 1 Audit Scanner per PolicyReport (or Audit Scanner queue/endpoint). This Audit Scanner is the sole owner of the creating/deleting results in the PolicyReport. Hence, we would have an Audit Scanner per namespace, matching their PolicyReport, and an Audit Scanner for the ClusterPolicyReport.

I think one audit scanner per namespace is too much, we can have one for all and then create queues or other sync mechanism per namespace

> I understood it as the Audit Scanner being the ones creating and POSTing the requests, and waiting for the return code of the requests (409 conflict already exists). If all dry-run requests to the Kube API server gave a return, the audit run is complete.
> 
> Alternatively, we could naively wait for PolicyServers to report back results. Each PolicyServer will return X results, one per policy that applies, for each 1 request. We have Y PolicyServers, so X*Y results per 1 requests.

we can wait until all kube API returns. However, PolicyServer **might not** return X results, one per policy that applies, for each 1 request. If another webhook rejects the request before the policy evaluation PolicyServer won't see any request. Someone might have [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/) deployed alongside Kubewarden, that would also interfere in the results